dask/layers.py

```diff
@@ -103,6 +105,165 @@ def __init__(
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def is_materialized(self):
+        return hasattr(self, "_cached_dict")
+
+    def get_output_keys(self):
+        return self.keys()  # FIXME! this implementation materializes the graph
+
+    def _get_block_slices(self, shape=None, blockdims=None, index=None):
+        """Get a list (for each dimension) of dicts{blocknum: slice()}"""
+        try:
+            block_slices = self._block_slices
+        except AttributeError:
+            block_slices = list(map(_slice_1d, shape, blockdims, index))
+            self._block_slices = block_slices
+        return block_slices
+
+    def _keys_to_output_parts(self, keys):
+        """Simple utility to convert keys to array chunk indices."""
+        parts = set()
+        for key in keys:
+            try:
+                _name, *_part = key
+            except ValueError:
+                continue
+            if _name != self.out_name:
+                continue
+            parts.add(tuple(_part))
+        return parts
+
+    def _input_parts(self):
+        """Simple utility to get input chunk indices."""
+        parts_in = set()
+        block_slices = self._get_block_slices(self.shape, self.blockdims, self.index)
+        _part = []
+        for i in block_slices:
+            _part += sorted(i.keys())
+        parts_in.add(tuple(_part))
+        return parts_in
+
+    def cull(self, keys, all_keys):
+        """Cull a SlicingLayer HighLevelGraph layer."""
+        parts_out = self._keys_to_output_parts(keys)
+        parts_in = self._input_parts()
+        culled_deps = self._cull_dependencies(
+            keys, parts_out=parts_out, parts_in=parts_in
+        )
+        if parts_out != self.parts_out:
+            culled_layer = self._cull(parts_out)
+            return culled_layer, culled_deps
+        else:
+            return self, culled_deps
+
+    def _cull(self, parts_out):
+        return SlicingLayer(
+            out_name=self.out_name,
+            in_name=self.in_name,
+            blockdims=self.blockdims,
+            new_blockdims=self.new_blockdims,
+            index=self.index,
+            shape=self.shape,
+            parts_out=parts_out,
+        )
+
+    def _cull_dependencies(self, keys, parts_out=None, parts_in=None):
+        """Determine the necessary dependencies to produce `keys`.
+
+        For simple slicing, output chunks depend on which areas are sliced.
+        This method does not require graph materialization.
+        """
+        deps = defaultdict(set)
+        parts_out = parts_out or self._keys_to_parts(keys)
```

Comment:
Since this code path isn't really used, we could also just make the arguments non-optional. In general I prefer non-nullable arguments unless there is a compelling reason for them